### ISSUE TITLE
PyQt5 standalone fails with `Incompatible library version` on macOS

### ISSUE THREAD

**Issue by melvyn2** [Nuitka/Nuitka#473]
```bash
$ python3 -m nuitka --version
0.6.6rc1
Python: 3.7.4 (default, Jul  9 2019, 18:13:23) 
Executable: /usr/local/opt/python/bin/python3.7
OS: Darwin
Arch: x86_64
```
Python was installed with homebrew, and nuitka with `pip install https://github.com/Nuitka/Nuitka/archive/develop.zip`.

Test example:
```python
import sys
from PyQt5 import QtWidgets

def main():
    app = QtWidgets.QApplication(sys.argv)
    window = QtWidgets.QMainWindow()
    button = QtWidgets.QPushButton("Hello, PyQt!")
    window.setCentralWidget(button)
    window.show()
    app.exec_()

if __name__ == '__main__':
    main()
```
The program was saved as tmp.py and compiled with `xcrun -sdk macosx python3.7 -m nuitka --standalone --follow-imports --plugin-enable=qt-plugins tmp.py`. `xcrun` was required to provide certain header files. Here is what happens when the executable is then ran:
```
$ ./tmp.dist/tmp
Traceback (most recent call last):
  File "/Users/melvyn/tmp/./tmp.dist/tmp.py", line 2, in <module>
ImportError: dlopen(/Users/melvyn/tmp/./tmp.dist/PyQt5/QtWidgets.so, 2): Library not loaded: @rpath/QtWidgets.framework/Versions/5/QtWidgets
  Referenced from: /Users/melvyn/tmp/./tmp.dist/PyQt5/QtWidgets.so
  Reason: Incompatible library version: QtWidgets.so requires version 5.13.0 or later, but QtWidgets provides version 5.7.0
```
Here is the full project folder: 
[tmp.zip](https://github.com/Nuitka/Nuitka/files/3443017/tmp.zip)

**Comment by kayhayen** [comment]
In your tmp.zip there is no QtWidgets file. Can you have a look at what `_detectBinaryPathDLLsMacOS` does in your case? It should resolve the @rpath and then Nuitka would include the file in the distribution folder.

Yours,
Kay

**Comment by melvyn2** [comment]
In `tmp.dist/PyQt5` there is a `QtWidgets.so` file. I'm a bit new to Nuitka, how should I run `_detectBinaryPathDLLsMacOS` (as in, what arguments to pass)?

**Comment by kayhayen** [comment]
Well, it's part of the Nuitka source code, which is Python, and you can edit it and look at what it does for you, that is what I meant.

**Comment by melvyn2** [comment]
Running `otool -L tmp.dist/PyQt5/*.so` reveals a bunch of missing dependencies:
```
tmp.dist/PyQt5/QtCore.so:
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.13.0, current version 5.13.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
tmp.dist/PyQt5/QtGui.so:
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.13.0, current version 5.13.0)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.13.0, current version 5.13.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
tmp.dist/PyQt5/QtWidgets.so:
	@rpath/QtWidgets.framework/Versions/5/QtWidgets (compatibility version 5.13.0, current version 5.13.0)
	@rpath/QtGui.framework/Versions/5/QtGui (compatibility version 5.13.0, current version 5.13.0)
	@rpath/QtCore.framework/Versions/5/QtCore (compatibility version 5.13.0, current version 5.13.0)
	/System/Library/Frameworks/DiskArbitration.framework/Versions/A/DiskArbitration (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/IOKit.framework/Versions/A/IOKit (compatibility version 1.0.0, current version 275.0.0)
	/System/Library/Frameworks/OpenGL.framework/Versions/A/OpenGL (compatibility version 1.0.0, current version 1.0.0)
	/System/Library/Frameworks/AGL.framework/Versions/A/AGL (compatibility version 1.0.0, current version 1.0.0)
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
tmp.dist/PyQt5/sip.so:
	/usr/lib/libc++.1.dylib (compatibility version 1.0.0, current version 400.9.4)
	/usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1252.250.1)
```
Copying the `PyQt5/Qt/lib` and `PyQt5/Qt/plugins` folders from the PyQt5 installation gets us further:
```
qt.qpa.plugin: Could not load the Qt platform plugin "cocoa" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: cocoa, minimal, offscreen, webgl.

Abort trap: 6
```
Running Nuitka without `--plugin-enable=qt-plugins` makes the program work, though not as intended. The current problems, then, are:
* PyQt frameworks are not getting copied
* `--plugin-enable=qt-plugins` not working

**Comment by kayhayen** [comment]
@melvyn2 The otool command is used by Nuitka to find out what to copy, and then later install_name_change or so to not make outside references in the dist folder.

Your mission is to tell, why the mentioned function doesn't detect these files to copy properly.

Yours,
Kay

**Comment by melvyn2** [comment]
The issue is that `_detectBinaryPathDLLsMacOS()` always resolves `@rpath` to the directory of the binary, instead of the actual `@rpath`(s). This means that, for example, the `QtCore` dependency resolves to `/usr/local/lib/python3.7/site-packages/PyQt5/QtCore.framework/Versions/5/QtCore` instead of `/usr/local/lib/python3.7/site-packages/PyQt5/Qt/lib/QtCore.framework/Versions/5/QtCore`. This causes `addDLLInfo()` to filter them out, as they don't exist at that location.

**Comment by melvyn2** [comment]
Now the question is, should a `@rpath` resolver be implemented, or should a qt-specific workaround be added?

**Comment by kayhayen** [comment]
The fix to the binary DLL path detection is of course the way to go, as it will apply everywhere, and will not require replicating knowledge.

I think the code we are discussing is this:

```
            if filename.startswith("@rpath/"):
                filename = os.path.join(original_dir, filename[7:])
            elif filename.startswith("@loader_path/"):
                filename = os.path.join(original_dir, filename[13:])
```

If you find anything in otool help that makes it output these ones resolved, it would be great. What is this, can you try it, @melvyn2 

```
       -L     Display the names and version numbers of  the  shared  libraries
              that  the  object file uses, as well as the shared library ID if
              the file is a shared library.
```

That code above should be considered our attempt at `@rpath` resolved. It's suspicous that it is using the same directory to reference against. I wonder now if for `@rpath`, this would then refer to `os.path.dirname(sys.executable)` and that should be working better maybe.

Yours,
Kay

**Comment by melvyn2** [comment]
I've added a `@rpath` resolver to my branch, (and fixed an issue where some dependencies were being dropped), but they are not being linked to, as `fixupBinaryDLLPaths()` feeds `install_name_tool` the resolved paths, when it expects the original ones.

**Comment by kayhayen** [comment]
Good news @melvyn2 of course I would love to see that as a PR.

As for `fixupBinaryDLLPaths` it is supposed to make the files after copy, no longer refer to outside of the distribution folder, for which `@executable_path` is used as a reference point. Is that not correct, and should only be that, with nothing appended. Not really sure what `dist_path` is there.

Yours,
Kay

**Comment by melvyn2** [comment]
What I mean, is that `fixupBinaryDLLPaths()` takes a mapping of original dependency paths to their paths in distributable. It feeds those paths to `install_name_tool`, which will only change the dependency paths if they are found in the binary. The problem is, paths that use `@rpath`wil have been resolved. 

An example: a binary requires `dependency.so`. It finds it by looking for `@rpath/lib/dependency.so`. During the freezing process, the `@rpath` will be resolved, and so Nuitka's reference for the dependency is now `/usr/lib/dependency.so`. When it calls `fixupBinaryDLLPaths()`, it will call `install_name_tool -change /usr/lib/dependency.so @executable_path/dependency.so binary`. The problem is, the binary only knows of `@rpath/lib/dependency.so`, *not* `/usr/lib/dependency.so`, so `install_name_tool` does nothing.

**Comment by kayhayen** [comment]
I see, I guess this never worked for the '@' flavor of values then, and we would need to keep track of the unresolved path, as well as the resolved path then. That sounds doable. 

I guess we need to make make `used_dlls` and `dll_map` retain the information. Something I could add I guess. Or you could try and do it with global variables, where the resolved path of a DLL is mapped back to the original path, but that won't be 1:1 in all cases, so that will be of limited success.

**Comment by kayhayen** [comment]
Can you make the PR then? I am hesitant to change the resolving code, if you already have a better one. Obviously it provides the input to be passed along @melvyn2

**Comment by melvyn2** [comment]
Okay, I will work on it.

**Comment by melvyn2** [comment]
The qt plugins seem to work fine, actually, except that it wasn't including they `styles` plugin. I've added it, and once it passes tests, I'll open a PR.

**Comment by kayhayen** [comment]
That's great, really looking forward to it. Very much appreciated.

You can control plugins from the command line, by listing them, "styles" is not yet in "sensible", because I couldn't make my mind up about dependencies, linking against all of KDE and Gnome seems not cool for deployment. But you can just use "sensible,styles" for now.

**Comment by melvyn2** [comment]
Okay, I'll roll back that commit.

**Comment by kayhayen** [comment]
@melvyn2 my suspect though is that for macOS and Windows it's pretty much OK, I doubt there is that many styles. Not sure how to force a specific style with PyQt5, but if we did that, Linux could also be good, but don't worry about that for now.

**Comment by melvyn2** [comment]
Fixed in #502

**Comment by kayhayen** [comment]
The release 0.6.8 was just made and contains the correction.
